### PR TITLE
Thread safety, misc fixes, and refactoring

### PR DESCRIPTION
**Pull request by rdelval**
* Changed thrift function call wrapper to be more explicitly named and to perform more safety checks.

* Moved Jitter function from realis to retry.

* API code is now more uniform and follows a certain template.

* Lock added whenever a thrift call is made or when a modification is done to the connection. Note that calling ReestablishConn externally may result in some race conditions. We will move to make this function private in the near future.

* CreateService now mimics CreateJob with Service=True more closely by launching all necessary instances at once.

* Revamped error handling with more consistency across the code base.

** Please take extra time reviewing as this PR involves a lot of changes to critical pieces. **
-----------------------------------------
## Please read instrucitons below ##

Before submitting, please make sure you run a vagrant box running Aurora with the latest version shown in .auroraversion and run go test from the project root.

To run an Aurora Vagrant image, follow the instructions here:
http://aurora.apache.org/documentation/latest/getting-started/vagrant/

* Have you run goformat on the project before submitting? yes

* Have you run go test on the project before submitting? Do all tests pass? yes, yes

* Does the Pull Request require a test to be added to the end to end tests? If so, has it been added? yes, yes

### CODE REVIEW

File: realis.go [hunk 1]
```diff
@@ -428,15 +418,51 @@ func basicAuth(username, password string) string {
...
 
+type auroraThriftCall func() (resp *aurora.Response, err error)
+
+// Takes a Thrift API function call and returns response and error.
+// If Error from the API call is retryable, the functions re-establishes the connection with Aurora by
+// using the same configuration used by the original client. Locks usage of and changes to client connection in order
+// to make realis sessions thread safe.
+func (r *realisClient) thriftCallHelper(auroraCall auroraThriftCall) (*aurora.Response, error) {
+	// Only allow one go-routine make use or modify the thrift client connection
+	r.lock.Lock()
```
Comment (kkrishna) [inline realis.go]: Use defer to Unlock?

Comment (rdelval) [reply realis.go]: If we do that then we unlock right before the function returns. I wanted to make the critical section as short as possible while still avoiding race conditions.

Comment (kkrishna) [reply realis.go]: defer guarantees execution no matter how or from where it runs.. panic scenario or not.. besides, rest of the logic after unlock is pretty light -- interpretation of response code.

Comment (rdelval) [reply realis.go]: It does allow someone using the library to recover with the lock released in that case. Not sure how much that really buys us but I'm fine with it as long as the size of this function doesn't increase too much down the line.

**Pull request merged**